src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs

```diff
@@ -2,150 +2,135 @@
+            [Import(typeof(IProjectGuidService))]IProjectGuidService2 projectGuidService,
+            IActiveConfiguredProjectSubscriptionService projectSubscriptionService,
             ActiveConfiguredProject<DebuggerLaunchProviders> launchProviders)
+            : this(AsyncServiceProvider.GlobalProvider, threadingService, projectGuidService, projectSubscriptionService, launchProviders)
         {
-            Requires.NotNull(startupProjectsListService, nameof(startupProjectsListService));
-            Requires.NotNull(threadingService, nameof(threadingService));
-            Requires.NotNull(activeConfiguredProjectSubscriptionService, nameof(activeConfiguredProjectSubscriptionService));
-            Requires.NotNull(launchProviders, nameof(launchProviders));
+        }
 
-            _startupProjectsListService = startupProjectsListService;
+        public StartupProjectRegistrar(
+            IAsyncServiceProvider serviceProvider,
+            IProjectThreadingService threadingService,
+            IProjectGuidService2 projectGuidService,
+            IActiveConfiguredProjectSubscriptionService projectSubscriptionService,
+            ActiveConfiguredProject<DebuggerLaunchProviders> launchProviders)
+        : base(threadingService.JoinableTaskContext)
+        {
+            _serviceProvider = serviceProvider;
             _threadingService = threadingService;
-            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+            _projectGuidService = projectGuidService;
+            _projectSubscriptionService = projectSubscriptionService;
             _launchProviders = launchProviders;
         }
 
         [ProjectAutoLoad(startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted)]
         [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
         internal Task Load()
         {
-            EnsureInitialized();
-            return Task.CompletedTask;
-        }
-
-        protected override void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                _evaluationSubscriptionLink?.Dispose();
-            }
+            return InitializeAsync();
         }
 
-        protected override void Initialize()
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
         {
-            var watchedEvaluationRules = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName);
-            var evaluationBlock = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(ConfigurationGeneralRuleBlock_ChangedAsync);
-            _evaluationSubscriptionLink = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock
-                                            .LinkTo(target: evaluationBlock, ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates: true);
-        }
+            _projectGuid = await _projectGuidService.GetProjectGuidAsync()
```

Comment:
We can assume the GUID never changes over the lifetime of a project one the factory has completed.